va/va_test.go

```diff
@@ -1117,3 +1118,261 @@ func TestParseResults(t *testing.T) {
+			fmt.Sprintf("wrong fallbackAddress, expected %#v, got %#v", tc.result, result))
+	}
+}
+
+func TestFallbackDialer(t *testing.T) {
+	// Create a test VA
+	va, _, _ := setup()
+
+	// Create a new challenge to use for the httpSrv
+	chall := core.HTTPChallenge01()
+	setChallengeToken(&chall, core.NewToken())
+
+	// Create an IPv4 test server
+	hs := httpSrv(t, chall.Token)
+	defer hs.Close()
+
+	// Figure out what port the test server is on, and configure the VA to use
+	// that for HTTP challenges
+	port, err := getPort(hs)
+	test.AssertNotError(t, err, "failed to get test server port")
+	va.httpPort = port
+
+	// Create an identifier for a host that has an IPv6 and an IPv4 address.
+	// Since the IPv6First feature flag is not enabled we expect that the IPv4
+	// address will be used and validation will succeed using the httpSrv we
+	// created earlier.
+	host := "ipv4.and.ipv6.localhost"
+	ident = core.AcmeIdentifier{Type: core.IdentifierDNS, Value: host}
+	fmt.Printf("Validating challenge with no IPv6 fallback\n")
+	records, prob := va.validateChallenge(ctx, ident, chall)
+	test.Assert(t, prob == nil, "validation failed for an dual homed host with IPv6First disabled")
+	// We expect one validation record to be present
+	test.AssertEquals(t, len(records), 1)
+	// We expect that the address used was the IPv4 address
+	test.AssertEquals(t, records[0].AddressUsed.String(), "127.0.0.1")
+	// We expect that zero addresses were tried before the address used
+	test.AssertEquals(t, len(records[0].AddressesTried), 0)
+
+	// Enable the IPv6 First feature
+	_ = features.Set(map[string]bool{"IPv6First": true})
+	defer features.Reset()
+
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	scope := mock_metrics.NewMockScope(ctrl)
+	va.stats = scope
+
+	// We expect the IPV4 Fallback stat to be incremented
+	scope.EXPECT().Inc("IPv4Fallback", int64(1)).Return(nil)
+
+	// The validation is expected to succeed with IPv6First enabled even though
+	// the V6 server doesn't exist because we fallback to the IPv4 address.
+	fmt.Printf("Validating challenge with IPv6 fallback\n")
+	records, prob = va.validateChallenge(ctx, ident, chall)
+	test.Assert(t, prob == nil, "validation failed with IPv6 fallback to IPv4")
+	// We expect one validation record to be present
+	test.AssertEquals(t, len(records), 1)
+	// We expect that the address used was the IPv4 localhost address
+	test.AssertEquals(t, records[0].AddressUsed.String(), "127.0.0.1")
+	// We expect that one address was tried before the address used
+	test.AssertEquals(t, len(records[0].AddressesTried), 1)
+	// We expect that IPv6 address was tried before the address used
+	test.AssertEquals(t, records[0].AddressesTried[0].String(), "::1")
+}
+
+func TestFallbackTLS(t *testing.T) {
+	// Create a test VA
+	va, _, _ := setup()
+
+	// Create a new challenge to use for the httpSrv
+	chall := createChallenge(core.ChallengeTypeTLSSNI01)
+	setChallengeToken(&chall, core.NewToken())
```

Comment:
Why set a specific token here rather than relying on the `expectedToken` that `createChallenge` sets by default?